2.0-proposal/readme.md

```diff
@@ -0,0 +1,183 @@
+
+### Our internal state has changed! [internal]
+
+The internal state of the unit has changed and we need to react to that.
+
+Ex: A file changing (`file.changed.<xyz>`), landscape messing about on the system (removal of `apt.installed.xyz`), a flag being toggled or the data of a flag being changed (`config.changed` unsets `config.checked`).
+
+<!-- Actions is a tricky one. I'd wait a bit to include actions in the story until we're more certain of where we want to go.
+
+### Action requested! [external]
+
+The operator has requested an action that shouldn't significantly impact the state of the application.
+
+Ex: `smoke-test`, `run-job`, `create-backup`, ... -->
+
+# When do triggers and handlers run?
+
+The flow of the framework is the following:
+
+1. The reactive framework will run a handler when all its preconditions are met. A handler that has run is inactive. It will not run again until one of the flags it reacts to has changed. This change can be either the flag being toggled, or the flag data changing.
+
+2. After each handler run, all internal events will be processed. The reactive framework will not remove flags itself. Only triggers and handlers can do that.
+
+<!-- We should do step 2. recursively. This makes it possible for charmers to create deadlocks/ infinite loops. It would be good if the reactive framework could notice an infinite loop and crash. -->
+
+<!-- Triggers or "Connected States" fix a number of issues:  -->
+<!-- multiple handlers can watch the same file for changes by creating a trigger on `file.changed.xyz` that flags `myhandler.file.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/25 https://github.com/juju-solutions/charms.reactive/issues/44 -->
+<!-- when_resource_changed is a trigger that flags `resource.changed.xyz` https://github.com/juju-solutions/charms.reactive/issues/87 -->
+
+
+3. After all triggers have been processed, the reactive framework will run the next handler whose preconditions are met. (back to 1.)
+
+The reactive framework goes into sleep mode when there isn't a handler to execute anymore at step 1. The reactive framework wakes up when an event happens.
+
+# What is the difference between external and internal events?
+
+Handlers run transactionally. Internal events are processed after a handler run. External events are processed when the reactive framework wakes up.
+
+ - **All** internal events are processed after **every** handler run.
+ - **Only one** external event is processed when the reactive framework wakes up. If there are still external events waiting to be processed at the moment the reactive framework goes to sleep, it will wake back up immediately and process the next external event.
+
+# Show me the code!
+
+
+```python
+# Register triggers
+charms.reactive.register_trigger(
+  event='config.changed',
+  unset='config.checked')
```

Comment:
It's more than just "unset a flag before any handlers run".

It's also a way for layers to use flags without having to worry about what upper or lower layers might do with those flags. With a trigger, a layer can say "**I want to run when the config has changed even if another handler already removed that flag.**" The layer creates the following trigger.

```python
trigger(
    event='config.changed'
    set='mylayer.config.changed')
```

That layer is then free to remove `mylayer.config.changed` whenever it wants without having to worry of suppressing `config.changed` for other layers.

The weird behavior of the `config.changed` flag in reactive 1.0 is solved by triggers. That's why I'm so exited about triggers. It fixes all the problems and inconsistencies we have in reactive 1.0. We don't need an extra `file_changed` or a `resource_changed` decorator. These can all be implemented using triggers with custom events. No need to put additional decorators in the reactive framework. Layer devs can fix their own problems. Less work for us and more freedom for layer devs to experiment. Who knows what kind of events they might find useful?

Another example is the behavior @stub42 wants for the apt layer; he wants to know if landscape removed a packages. The apt packages changing becomes a custom event that triggers a flag.

I haven't elaborated on custom events, maybe I should also write a draft for that.